File: javascript/leetcode/next-greater-element-ii.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var nextGreaterElements = function(nums) {
	let result = Array(nums.length);
	let stack = [];
	for (var i = (nums.length * 2) - 1; i >= 0; i--) {
		while (stack.length > 0 && nums[stack[stack.length - 1]] <= nums[i % nums.length]) {
			stack.pop();
		}
		result[i % nums.length] = stack.length == 0 ? -1 : nums[stack[stack.length - 1]];
		stack.push(i % nums.length);
	}
	return result;
};

File: python/leetcode/next-greater-element-ii.py
class Solution:
    def scanForGreater(self, target, idx, nums):
        i = 0
        l = len(nums)
        while i < l:
            if nums[(idx + i) % l] > target:
                return nums[(idx + i) % l]
            i += 1
        return -1
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        result = []
        for idx in range(len(nums)):
            n = nums[idx]
            result.append(self.scanForGreater(n, idx, nums))
        return result
